File: prodigy_recipes/ocr_eval.js
// Shared state that should persist across multiple tasks
const STATE = {
  currentTask: null,
  pickerOpen: false,
  inputInitialFontSize: null,
}

function hookIntoDOM() {
  const input = document.getElementById("transcription");
  if (!document.getElementById('transcription-warnings')) {
    const container = input.parentElement.parentElement;
    container.insertAdjacentHTML(
      'afterend',
      `<div class="prodigy-content"
            style="padding: 0 20px 20px 20px; margin-top: 0;">
        <div id="transcription-warnings"></div>
      </div>`)
  }
  input.addEventListener('input', evt => {
    adjustInputFontSize();
    highlightSpecialCharacters();
  });
  input.addEventListener('change', evt => {
    adjustInputFontSize();
    highlightSpecialCharacters();
  });

  // Update input font size and background when input is initialized
  adjustInputFontSize();
  highlightSpecialCharacters();

  // Update viewer link
  const viewerLink = document.getElementById('viewer-anchor');
  const { viewer_url } = window.prodigy.content;
  viewerLink.href = viewer_url;

}

// Redirect user to a new session
function redirectToSession(session) {
  const currentUrl = new URL(window.location.href);
  currentUrl.searchParams.set('session', session);
  window.location.href = currentUrl.href;
}

// Register event callbacks for input
document.addEventListener('prodigymount', evt => {
  // Set initial task in state
  STATE.currentTask = window.prodigy.content._task_hash;

  // Update input font size and background when input changes
  const input = document.getElementById("transcription");
  const style = window.getComputedStyle(input, null);
  STATE.inputInitialFontSize = Number.parseInt(style.getPropertyValue('font-size').slice(0, -2), 10);
  hookIntoDOM();
});

// If the session is launched without a named session, ask the user to select one
// of the sessions defined with the PRODIGY_POSSIBLE_SESSIONS environment variable.
// The selection can optionally (default: enabled) be remembered in localStorage, in
// this case anonymous future sessions will be redirected to the saved session.
document.addEventListener('prodigymount', evt => {
  const currentSession = window.prodigy.config.session;
  const possibleSessions = __allowed_named_sessions__;  // Will be templated in
  if (currentSession && possibleSessions.includes(currentSession)) {
    return;
  }
  const userSession = localStorage.getItem('prodigy_default_session');
  if (userSession) {
    redirectToSession(userSession);
    return;
  }

  let rememberSession = true;
  let selectedSession = possibleSessions[0];
  const tmpl = document.createElement('template');
  tmpl.innerHTML = `
  <div style="display: flex; position: fixed; top: 0px; left: 0px; width: 100%; height: 100%; background-color: rgba(0, 0, 0, 0.8); z-index: 9999; align-items: center; justify-content: center;">
    <div style="width: 50%; background-color: white; display: flex; align-items: center; flex-direction: column; font-family: sans-serif; padding: 1em; border-radius: 1.5em;">
      <p style="font-weight: bold; font-size: 3em;">
        Bitte einen Benutzer auswählen:
      </p>
      <select style="margin-top: 2em; font-size: 2em; padding: 0.5em;">
        ${possibleSessions.map(sess => `<option value="${sess}">${sess}</option>`).join('\n')}
      </select>
      <label style="display: block; font-size: 1.5em; margin-top: 1em;">
        <input type="checkbox" style="margin-right: 1em;" checked="true">
        Benutzer merken
      </label>
      <button style="font-size: 2em; padding: 0.5em; margin-top: 1em; background-color: lightgray;">
        Nutzer-Session starten
      </button>
    </div>
  </div>
  `.trim();
  tmpl.content.querySelector('select').addEventListener('change', evt => {
    selectedSession = evt.target.value;
  });
  tmpl.content.querySelector('input').addEventListener('change', evt => {
    rememberSession = evt.target.checked;
  });
  tmpl.content.querySelector('button').addEventListener('click', evt => {
    if (rememberSession) {
      localStorage.setItem('prodigy_default_session', selectedSession);
    }
    redirectToSession(selectedSession)
  });

  document.body.appendChild(tmpl.content.firstChild);
});

// When the task changes, synchronize the character picker state in the DOM,
// update input box styling and register event handlers so input style remains
// in sync with input value.
document.addEventListener('prodigyupdate', evt => {
  const task = evt.detail.task._task_hash;
  if (task === STATE.currentTask) {
    // No change
    return;
  }
  STATE.currentTask = task;

  // Keep picker open across multiple tasks once it has been opened
  const picker = document.querySelector(".character-picker");
  if (STATE.pickerOpen) {
    picker.style.display = "grid";
  } else if (!STATE.pickerOpen) {
    picker.style.display = "none";
  }

  hookIntoDOM();
});

function toggleCharacterPicker() {
  const picker = document.querySelector(".character-picker");
  if (picker.style.display === "none") {
    picker.style.display = "grid";
    STATE.pickerOpen = true;
  } else {
    picker.style.display = "none";
    STATE.pickerOpen = false;
  }
}

// React hijacks the input event, so we need to do some special stuff to get our value
// to sync with the React state.
function setValueForReact(inputElem, value) {
  const lastVal = inputElem.value;
  inputElem.value = value;
  const evt = new Event('input', { target: inputElem, bubbles: true });
  evt.simulated = true;
  if (inputElem._valueTracker) {
    inputElem._valueTracker.setValue(lastVal);
  }
  inputElem.dispatchEvent(evt);
}


function insertGrapheme(grapheme) {
  const input = document.getElementById("transcription");
  const start = input.selectionStart ?? 0;
  const end = input.selectionEnd ?? 0;
  const text = input.value;

  setValueForReact(input, text.slice(0, start) + grapheme + text.slice(end));

  input.focus();
  input.setSelectionRange(start + grapheme.length, start + grapheme.length);
}

function insertUncertaintyMarker() {
  const input = document.getElementById("transcription");
  const start = input.selectionStart ?? 0;
  const end = input.selectionEnd ?? 0;
  // Only works for selections of one or more characters
  if (start === end) {
    return;
  }
  const marked = `⟅${input.value.slice(start, end)}⟆`;

  setValueForReact(input, input.value.slice(0, start) + marked + input.value.slice(end));

  input.focus();
  input.setSelectionRange(start + marked.length, start + marked.length);
}

function getInputCharacterCoordinates() {
  const input = document.getElementById("transcription");
  const styles = window.getComputedStyle(input, null);
  if (!getInputCharacterCoordinates._canvas) {
    getInputCharacterCoordinates._canvas = document.createElement('canvas');
  }
  const canvas = getInputCharacterCoordinates._canvas;
  const ctx = canvas.getContext('2d');
  function _getStyle(key, fallback) {
    return styles.getPropertyValue(key) || fallback;
  }
  ctx.font = `${_getStyle('font-weight', 400)} ${_getStyle('font-size', '20px')} ${_getStyle('font-family', 'monospace')}`;
  const out = [];
  let offset = Number.parseInt(_getStyle('padding-left', '0px').slice(0, -2), 10);
  for (let i = 0; i < input.value.length; i++) {
    const char = input.value[i];
    let width = ctx.measureText(char).width;
    if (char === '\uD83E' || char === '\uDD14') {
      width = width / 2;
    }
    out.push({ char, offset, width });
    offset += width;
  }
  return out;
}

function highlightSpecialCharacters() {
  const input = document.getElementById("transcription");
  // CSS linear color stops in the form `${color} ${start}px ${end}px`
  const gradientStops = [];
  const coords = getInputCharacterCoordinates();
  let lastOffset = 0;
  let hasS = false;
  let hasUmlauts = lastOffset;
  for (let i = 0; i < coords.length; i++) {
    let color;
    let { char, offset, width } = coords[i];
    if (char === 'ſ' || char === 's') {
      color = 'rgba(255, 0, 0, 0.5)';
      hasS = true
    } else if (char === 'f') {
      color = 'rgba(255, 0, 0, 0.1)';
    } else if ('äöüÄÖÜ'.indexOf(char) >= 0) {
      color = 'rgba(0, 128, 0, 0.5)';
      hasUmlauts = true;
    } else if (char ==='ͤ') {
      color = 'rgba(0, 128, 0, 0.5)';
      hasUmlauts = true;
      // Combining codepoint has no width, uses that of the previous character
      offset = coords[i-1].offset;
      width = coords[i-1].width;
    }
    if (char === '⟅') {
      // Mark the whole uncertainty range
      while (i < coords.length && coords[i].char !== '⟆') {
        i++;
        width += coords[i].width;
      }
      color = 'rgba(255, 255, 0, 0.5)';
    }
    // Update gradient stops:
    if (color) {
      if (offset !== lastOffset) {
        gradientStops.push(`rgba(0, 0, 0, 0) ${lastOffset}px ${offset}px`);
      }
      gradientStops.push(`${color} ${offset}px ${offset + width}px`);
      lastOffset = offset + width;
    }
  }
  gradientStops.push(`rgba(0, 0, 0, 0) ${lastOffset}px 100%`);
  input.style.background = `linear-gradient(to right, ${gradientStops.join(', ')})`;
  const warnings = [];
  if (hasS) {
    document.getElementById("fix-long-s-btn").disabled = false;
    warnings.push(['rgb(255, 0, 0)', 'Im Text kommen Formen des kleinen "s" vor!']);
  } else {
    document.getElementById("fix-long-s-btn").disabled = true;
  }
  if (hasUmlauts) {
    document.getElementById("fix-umlauts-btn").disabled = false;
    warnings.push(['rgb(0, 128, 0)', 'Im Text kommen Umlaute vor!']);
  } else {
    document.getElementById("fix-umlauts-btn").disabled = true;
  }
  const warningContainer = document.getElementById("transcription-warnings");
  warningContainer.innerHTML = "";
  if (warnings.length) {
    for (const [color, msg] of warnings) {
      warningContainer.insertAdjacentHTML('afterbegin', `<p style="color: ${color}" class="transcription-warning"><strong>⚠️ Achtung:</strong> ${msg}</p>`);
    }
  }
}

function fixLongS() {
  const input = document.getElementById("transcription");
  const text = input.value;
  const fixed = text.replace(/s/g, 'ſ');
  setValueForReact(input, fixed);
  highlightSpecialCharacters();
}

function fixUmlauts() {
  const input = document.getElementById("transcription");
  const text = input.value;
  const fixed = text.replace(/([äöüÄÖÜ])/g, (match, char) => {
    switch (char) {
      case 'ä': return 'aͤ';
      case 'ö': return 'oͤ';
      case 'ü': return 'uͤ';
      case 'Ä': return 'Aͤ';
      case 'Ö': return 'Oͤ';
      case 'Ü': return 'Uͤ';
      default: return match;
    }
  });
  setValueForReact(input, fixed);
  adjustInputFontSize();
  highlightSpecialCharacters();
}

function adjustInputFontSize() {
  const input = document.getElementById("transcription");
  const style = window.getComputedStyle(input, null);
  const paddingLeft = Number.parseInt(style.getPropertyValue('padding-left').slice(0, -2), 10);
  const paddingRight = Number.parseInt(style.getPropertyValue('padding-right').slice(0, -2), 10);
  const maxWidth = input.clientWidth - paddingLeft - paddingRight;
  if (input.scrollWidth == maxWidth) {
    input.style.fontSize = `${STATE.inputInitialFontSize}px`;
    return;
  }
  const text = input.value;
  let fontSize = Number.parseInt(style.getPropertyValue('font-size').slice(0, -2), 10);
  const fontFamily = style.getPropertyValue('font-family') || 'monospace';
  const fontWeight = style.getPropertyValue('font-weight') || 'normal';
  const ctx = document.createElement('canvas').getContext('2d');
  ctx.font = `${fontWeight} ${fontSize}px ${fontFamily}`;
  let textWidth = ctx.measureText(text).width;
  while (textWidth > maxWidth) {
    fontSize -= 0.1;
    ctx.font = `${fontWeight} ${fontSize}px ${fontFamily}`;
    textWidth = ctx.measureText(text).width;
  }
  input.style.fontSize = `${fontSize}px`;
}

function showFullScreenLineModal() {
  const { line_image_url } = window.prodigy.content;
  document.body.insertAdjacentHTML('beforeend', `
    <div id="line-modal" style="display: flex; position: fixed; top: 0px; left: 0px; width: 100%; height: 100%; background-color: rgba(0, 0, 0, 0.8); z-index: 9999; align-items: center; justify-content: center;">
      <div style="width: 100%; background-color: white; display: flex; align-items: center; flex-direction: column; font-family: sans-serif; padding: 0.5em; border-radius: 1.5em;">
        <img src="${line_image_url}" style="width: 100%; max-height: 80vh; object-fit: contain;"/>
      </div>
    </div>
  `);
  document.getElementById('line-modal').addEventListener('click', closeFullScreenLineModal);
}

function closeFullScreenLineModal() {
  document.getElementById("line-modal").remove();
}

File: prodigy_recipes/ocr_eval.py
import json
import os
import re
import unicodedata
from pathlib import Path

import prodigy
from prodigy.components.loaders import JSONL
from prodigy.util import get_config


def get_unicode_name(grapheme: str) -> str:
    return " + ".join(unicodedata.name(c).title() for c in grapheme)


UNCERTAINTY_DELIMITER_PAT = re.compile(r"🤔?⟅(.+?)⟆")
config = get_config()
SPECIAL_CHARACTERS = (
    config["keyboard"]
    if "keyboard" in config
    else [
        "ſ",
        "ā",
        "ē",
        "ī",
        "ō",
        "ū",
        "n̄",
        "m̄",
        "p̄",
        "t̄",
        "q̄",
        "q̈",
        "Ā",
        "Ē",
        "Ī",
        "Ō",
        "Ū",
        "aͤ",
        "oͤ",
        "uͤ",
        "Aͤ",
        "Oͤ",
        "Uͤ",
        "ꝗ",
        "q́",
        "å",
        "ů",
        "ꝛ",
        "⁊",
        "Æ",
        "Œ",
        "æ",
        "œ",
        "ë",
        "ↄ",
        "ę",
        "｢",
        "｣",
        "⁰",
        "¹",
        "²",
        "³",
        "⁴",
        "⁵",
        "⁶",
        "⁷",
        "⁸",
        "⁹",
        "⸗",
        "—",
        "‹",
        "›",
        "»",
        "«",
        "„",
        "”",
        "’",
        "£",
        "§",
        "†",
        "½",
        "¼",
        "¾",
        "⅓",
        "⅔",
        "⅕",
        "⅖",
        "⅗",
        "⅘",
        "⅙",
        "⅐",
        "⅚",
        "⅛",
        "⅜",
        "⅝",
        "⅞",
        "⅑",
        "⅒",
        "Α",
        "Δ",
        "Κ",
        "Π",
        "Σ",
        "ά",
        "έ",
        "ή",
        "ί",
        "α",
        "β",
        "γ",
        "δ",
        "ε",
        "ζ",
        "η",
        "θ",
        "ι",
        "κ",
        "λ",
        "μ",
        "ν",
        "ξ",
        "ο",
        "π",
        "ρ",
        "ς",
        "σ",
        "τ",
        "υ",
        "φ",
        "χ",
        "ψ",
        "ω",
        "ό",
        "ύ",
        "ώ",
        "ϑ",
        "ϰ",
        "ϱ",
    ]
)
GLOBAL_CSS = """
.prodigy-container {
    width: fit-content;
}

input#transcription {
    font-family: monospace;
}

#transcription-warnings {
    margin-top: -1em;
}

.transcription-warning {
    font-size: 1.25em;
    text-align: left;
}

.character-picker {
    text-aligns: center;
    font-size: 1.25em;
    height: 136px;
    overflow-y: scroll;
    display: grid;
    grid-template-columns: repeat(auto-fit, minmax(54px, 1fr));
    grid-template-rows: repeat(auto-fit, minmax(54px, 1fr));
    grid-gap: 4px;
    justify-content: center;
}

.character-picker .character {
  all: unset;
  display:inline-block;
  width:50px;
  height:50px;
  margin: 3px;
  background: #fff;
  border-radius: 4px;
  box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.5);
  font: 30px/50px 'Lato', "Trebuchet MS", Roboto, Helvetica, Arial, sans-serif;
  text-align: center;
  color: #444;
}

.character-picker .character:hover {
    background: #eee;
    cursor: pointer;
}

.character-picker .character:active {
    background: #ddd;
}

.btn-container {
    display: flex;
    justify-content: space-between;
    margin-bottom: -1em;
}

.btn-container button {
    padding: 8px !important;
    text-transform: none !important;
}

.btn-container button:disabled {
    opacity: 0.5;
    cursor: default;
}

.prodigy-container {
    padding-top: 2em;
}

.context-container {
    display: flex;
    flex-direction: column;
    line-height: 1 !important;
    white-space: normal;
}

.prodigy-content {
    margin-top: -3em;
}

.viewer-link {
    width: 100%;
    font-size: 0.5em;
    margin-top: 0.25em;
    margin-bottom: 0.25em !important;
    text-align: left;
}
"""
HTML_TEMPLATE = f"""
<div class="context-container">
    <p class="viewer-link">
        <a id="viewer-anchor"
           href=""
           target="_blank">Seite im Viewer öffnen</a>
    </p>
    <svg viewBox="0 0 {{{{context_area.width}}}} {{{{context_area.height}}}}" style="width: min(100%, {{{{context_area.width}}}}px)">
        <defs>
            <mask id="highlight-mask">
                <!-- transparent shape that matches the area to highlight -->
                <rect x="0" y="0" width="{{{{context_area.width}}}}" height="{{{{context_area.height}}}}" fill="white" />
                <rect x="{{{{line_area.x}}}}"
                    y="{{{{line_area.y}}}}"
                    width="{{{{line_area.width}}}}"
                    height="{{{{line_area.height}}}}"
                    fill="black" />
            </mask>
        </defs>
        <a onclick="showFullScreenLineModal()" style="cursor: pointer;">
            <image xlink:href="{{{{context_image_url}}}}" />
            <rect x="0"
                y="0"
                width="{{{{context_area.width}}}}"
                height="{{{{context_area.height}}}}"
                mask="url(#highlight-mask)"
                fill="black"
                opacity="0.5" />
        </a>
    </svg>
</div>
<div class="btn-container">
    <button class="keyboard-toggle" onclick="toggleCharacterPicker()" title="Hilfstastatur">
        <svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 576 512"><!--! Font Awesome Pro 6.4.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license (Commercial License) Copyright 2023 Fonticons, Inc. --><path d="M64 64C28.7 64 0 92.7 0 128V384c0 35.3 28.7 64 64 64H512c35.3 0 64-28.7 64-64V128c0-35.3-28.7-64-64-64H64zm16 64h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V144c0-8.8 7.2-16 16-16zM64 240c0-8.8 7.2-16 16-16h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V240zm16 80h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H80c-8.8 0-16-7.2-16-16V336c0-8.8 7.2-16 16-16zm80-176c0-8.8 7.2-16 16-16h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H176c-8.8 0-16-7.2-16-16V144zm16 80h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H176c-8.8 0-16-7.2-16-16V240c0-8.8 7.2-16 16-16zM160 336c0-8.8 7.2-16 16-16H400c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H176c-8.8 0-16-7.2-16-16V336zM272 128h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H272c-8.8 0-16-7.2-16-16V144c0-8.8 7.2-16 16-16zM256 240c0-8.8 7.2-16 16-16h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H272c-8.8 0-16-7.2-16-16V240zM368 128h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H368c-8.8 0-16-7.2-16-16V144c0-8.8 7.2-16 16-16zM352 240c0-8.8 7.2-16 16-16h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H368c-8.8 0-16-7.2-16-16V240zM464 128h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H464c-8.8 0-16-7.2-16-16V144c0-8.8 7.2-16 16-16zM448 240c0-8.8 7.2-16 16-16h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H464c-8.8 0-16-7.2-16-16V240zm16 80h32c8.8 0 16 7.2 16 16v32c0 8.8-7.2 16-16 16H464c-8.8 0-16-7.2-16-16V336c0-8.8 7.2-16 16-16z"/></svg>
    </button>
    <button id="fix-long-s-btn" onclick="fixLongS()" title="Alle 's' zu 'ſ' korrigieren" disabled>
        s→ſ
     </button>
     <button id="fix-umlauts-btn" onclick="fixUmlauts()" title="Alle Umlaute zu historischen Formen korrigieren" disabled>
        ä→aͤ
     </button>
    <button class="uncertainty-btn"
            onclick="insertUncertaintyMarker()"
            title="Selektion als unsicher markieren">
        🤔
    </button>
</div>
<div class="character-picker" style="display: none;">
{''.join(f'<button onclick="insertGrapheme(this.dataset.char)" class="character" title="{get_unicode_name(c)}" data-char="{c}">{c}</button>' for c in SPECIAL_CHARACTERS)}
</div>
"""


def before_db(examples: list[dict]) -> list[dict]:
    """Unify uncertainty markers to their original format.

    We decided to not show the emoji in the UI, since it breaks our input background
    highlighting approach. Instead, we add the emoji before saving to the database.
    """
    for ex in examples:
        # Remove, since this is just a dummy blank string
        del ex["html"]
        ex["transcription"] = UNCERTAINTY_DELIMITER_PAT.sub(
            r"🤔⟅\1⟆", ex["transcription"], count=0
        )
    return examples


@prodigy.recipe("ocr-eval")
def ocr_eval(dataset: str, tasks_jsonl: str):
    def get_stream():
        src = JSONL(tasks_jsonl)
        for task in src:
            line_hash = "-".join(
                (
                    task["volume_id"],
                    f"{task['page_num']:05d}",
                    str(task["context_area"]["x"] + task["line_area"]["x"]),
                    str(task["context_area"]["y"] + task["line_area"]["y"]),
                    str(task["line_area"]["width"]),
                    str(task["line_area"]["height"]),
                )
            )
            yield {**task, "html": " ", "line_hash": line_hash}

    return {
        "view_id": "blocks",
        "dataset": dataset,
        "stream": (
            prodigy.set_hashes(ex, input_keys={"line_hash"}) for ex in get_stream()
        ),
        "before_db": before_db,
        "config": {
            "blocks": [
                {
                    "view_id": "html",
                    "html_template": HTML_TEMPLATE,
                },
                {
                    "view_id": "text_input",
                    "field_id": "transcription",
                    "field_label": "Transkription",
                    "field_rows": 1,
                    "field_autofocus": False,
                },
            ],
            "javascript": (Path(__file__).parent / "./ocr_eval.js")
            .read_text()
            .replace(
                "__allowed_named_sessions__",
                json.dumps(os.environ.get("PRODIGY_POSSIBLE_SESSIONS", "").split(",")),
            ),
            "global_css": GLOBAL_CSS,
        },
    }
